runtime-modules/storage/src/lib.rs

```diff
@@ -3942,4 +3602,342 @@ impl<T: Trait> Module<T> {
             distribution_bucket_index,
         }
     }
+
+    fn generate_new_storage_buckets(
+        bucket_ids: &BTreeSet<T::StorageBucketId>,
+        new_voucher_update: VoucherUpdate,
+        op: &BagOperationParams<T>,
+    ) -> Result<BTreeMap<T::StorageBucketId, StorageBucket<T>>, Error<T>> {
+        // create temporary iterator of bucket
+        let tmp = bucket_ids
+            .iter()
+            // discard non existing bucket and build a (bucket_id, bucket) map
+            .filter_map(|id| {
+                Self::ensure_storage_bucket_exists(&id)
+                    .map(|bk| (*id, bk))
+                    .ok()
+            })
+            // attempt to update bucket voucher
+            .map(|(id, bk)| {
+                bk.voucher
+                    .clone()
+                    .update(new_voucher_update)
+                    .map_or_else(
+                        |e| match e {
+                            "size" => Err(Error::<T>::StorageBucketObjectSizeLimitReached),
+                            _ => Err(Error::<T>::StorageBucketObjectNumberLimitReached),
+                        },
+                        |voucher| Ok((id, voucher)),
+                    )
+                    .map(|(id, voucher)| (id, StorageBucket::<T> { voucher, ..bk }))
+            });
+
+        match op {
+            BagOperationParams::<T>::Create(..) => {
+                let correct = tmp
+                    .filter_map(|r| {
+                        r.map(|(id, mut bk)| {
+                            bk.register_bag_assignment();
+                            (id, bk)
+                        })
+                        .ok()
+                    })
+                    .collect::<BTreeMap<_, _>>();
+                ensure!(
+                    !correct.is_empty(),
+                    Error::<T>::StorageBucketIdCollectionsAreEmpty
+                );
+                Ok(correct)
+            }
+            BagOperationParams::<T>::Update(..) => tmp.collect(),
+            BagOperationParams::<T>::Delete => tmp
+                .map(|r| {
+                    r.map(|(id, mut bk)| {
+                        bk.unregister_bag_assignment();
+                        (id, bk)
+                    })
+                })
+                .collect(),
+        }
+    }
+
+    fn generate_new_distribution_buckets(
```

Comment:
The name is confusing - we don't create new buckets here.

Reply:
:+1: